in_toto/runlib.py

```diff
@@ -119,28 +133,56 @@ def record_artifacts_as_dict(artifacts):
   if not artifacts:
     return artifacts_dict
 
-  # Exclude excluded files and dirs from the first level of tree traversel
-  for artifact in _apply_exclude_patterns(artifacts,
+  # Normalize passed paths
+  norm_artifacts = []
+  for path in artifacts:
+    norm_artifacts.append(os.path.normpath(path))
+
+  # Iterate over remaining normalized artifact paths after
+  # having applied exclusion patterns
+  for artifact in _apply_exclude_patterns(norm_artifacts,
         settings.ARTIFACT_EXCLUDES):
 
     if not os.path.exists(artifact):
       log.warning("path: {} does not exist, skipping..".format(artifact))
       continue
 
     if os.path.isfile(artifact):
-      artifacts_dict[_normalize_path(artifact)] = _hash_artifact(artifact)
+      # Path was already normalized above
+      artifacts_dict[artifact] = _hash_artifact(artifact)
 
     elif os.path.isdir(artifact):
       for root, dirs, files in os.walk(artifact):
-        # Also exclude excluded subdirs
-        # Modifying the dirs variable makes walk only recurse into
-        # remaining dirs
-        dirs[:] = _apply_exclude_patterns(dirs, settings.ARTIFACT_EXCLUDES)
-
-        # Finally exclude excluded tree nodes (files)
-        for name in _apply_exclude_patterns(files, settings.ARTIFACT_EXCLUDES):
-          filepath = os.path.join(root, name)
-          artifacts_dict[_normalize_path(filepath)] = _hash_artifact(filepath)
+
+        # Create a list of normalized dirpaths
+        dirpaths = []
+        for dirname in dirs:
+          norm_dirpath = os.path.normpath(os.path.join(root, dirname))
+          dirpaths.append(norm_dirpath)
+
+        # Apply exlcude patterns on normalized dirpaths
+        dirpaths = _apply_exclude_patterns(dirpaths, settings.ARTIFACT_EXCLUDES)
+
+        # Reset and refill dirs with remaining names after exclusion
+        # Modify (not reassign) dirnames to only recurse into remaining dirs
+        dirs[:] = []
```

Comment:
From the python [`os.walk docs`](https://docs.python.org/2/library/os.html#os.walk)"
> ...the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames...